src/extension/declarative/signature.rs

```diff
@@ -0,0 +1,257 @@
+pub(super) struct SignatureDeclaration {
+    /// The inputs to the operation.
+    inputs: Vec<SignaturePortDeclaration>,
+    /// The outputs of the operation.
+    outputs: Vec<SignaturePortDeclaration>,
+    /// A set of extensions invoked while running this operation.
+    #[serde(default)]
+    #[serde(skip_serializing_if = "crate::utils::is_default")]
+    extensions: ExtensionSet,
+}
+
+impl SignatureDeclaration {
+    pub fn make_signature(
+        &self,
+        ext: &Extension,
+        scope: &ExtensionSet,
+        registry: &ExtensionRegistry,
+        op_params: &[TypeParam],
+    ) -> Result<SignatureFunc, ExtensionDeclarationError> {
+        let body = FunctionType {
+            input: self.make_type_row(&self.inputs, ext, scope, registry, op_params)?,
+            output: self.make_type_row(&self.outputs, ext, scope, registry, op_params)?,
+            extension_reqs: self.extensions.clone(),
+        };
+
+        let poly_func = PolyFuncType::new(op_params, body);
+        Ok(SignatureFunc::TypeScheme(CustomValidator::from_polyfunc(
+            poly_func,
+        )))
+    }
+
+    /// Create a type row from a list of port declarations.
+    fn make_type_row(
+        &self,
+        v: &[SignaturePortDeclaration],
+        ext: &Extension,
+        scope: &ExtensionSet,
+        registry: &ExtensionRegistry,
+        op_params: &[TypeParam],
+    ) -> Result<TypeRow, ExtensionDeclarationError> {
+        let types = v
+            .iter()
+            .map(|port_decl| port_decl.make_types(ext, scope, registry, op_params))
+            .flatten_ok()
+            .collect::<Result<Vec<Type>, _>>()?;
+        Ok(types.into())
+    }
+}
+
+/// A declarative definition for a number of ports in a signature's input or output.
+///
+/// Serialized as a 2 or 3-element list, where:
+/// - The first element is an optional human-readable name of the port.
+/// - The second element is the port type id.
+/// - The optional third element is either:
+///     - A number, indicating a repetition of the port that amount of times.
+///     - A parameter identifier, to use as the repetition number.
+///     - Nothing, in which case we default to a single repetition.
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+struct SignaturePortDeclaration {
+    /// The description of the ports. May be `null`.
+    description: Option<String>,
+    /// The type for the port.
+    type_decl: TypeDeclaration,
+    /// The number of repetitions for this port definition.
+    repeat: PortRepetitionDeclaration,
+}
+
+impl SignaturePortDeclaration {
+    /// Return an iterator with the types for this port declaration.
+    ///
+    /// Only a fixed number of repetitions is supported for now.
+    ///
+    /// TODO: We may need to use custom signature functions if `repeat` depends on a variable.
+    fn make_types(
+        &self,
+        ext: &Extension,
+        scope: &ExtensionSet,
+        registry: &ExtensionRegistry,
```

Comment:
these parameters are passed in to things a lot - maybe worth defining a struct
to hold them? `DeclarationContext` perhaps?